### ISSUE TITLE
AudioParam setup

### ISSUE THREAD

**Issue by orottier** [ircam-ismm/node-web-audio-api#1]
I have no experience whatsoever with napi-rs, but as I understand your issue is probably that you need to return an owned object representing the `AudioParam` but the web-audio-api-rs only allows to retrieve an `&AudioParam` reference from an `AudioNode`? The `ParamGetter` acts as an owned object delegating access to the param?

I think your setup is not bad per se, I do have some suggestions:

Wrapping nodes inside an Rc is a very good idea since JS assumes shared ownership anyway. As you probably noticed we don't need `Rc<RefCell<..>>` as all functions on AudioNode take an `&`-binding, not `&mut`. I would suggest you use `Arc` instead `Rc` since JS is inherently multithreaded.

Instead of the unwieldy ParamGetter you have now:
```
pub(crate) enum ParamGetter {
  GainNodeGain(Rc<GainNode>),
  OscillatorNodeFrequency(Rc<OscillatorNode>),
}
```
You could use some generics:
```
pub(crate) struct ParamGetter<F> {
    audio_node: Rc<dyn AudioNode>,
    get_param_fn: F,
}
impl<F: Fn(&dyn AudioNode) -> &AudioParam> ParamGetter<F> {
    fn new(audio_node: Rc<dyn AudioNode>, get_param_fn: F) -> Self {
        Self { audio_node, get_param_fn }
    }

    fn get_param(&self) -> &AudioParam {
        (self.get_param_fn)(self.audio_node.as_ref())
    }
}

// use it like this:
let gain_node = .....;
let getter = ParamGetter::new(gain_node, |&gain_node| gain_node.gain());
```

This is completely untested so you may run into issues I did not foresee. But I hope it helps so you dont need to enumerate all the `GainNodeGain` and friends ;)

**Comment by orottier** [comment]
Ah, this gives the error:
```
error[E0599]: no method named `gain` found for reference `&dyn AudioNode` in the current scope
  --> src/gain_node.rs:49:61
   |
49 |   let param_getter2 = ParamGetter2::new(native_clone, |n| n.gain());
   |                                                             ^^^^ method not found in `&dyn AudioNode`
```

Nothing we can't fix, we need to be general over the incoming AudioNode too:
```
pub(crate) struct ParamGetter2<A, F> {
    audio_node: Rc<A>,
    get_param_fn: F,
}
impl<A: AudioNode, F: Fn(&A) -> &AudioParam> ParamGetter2<A, F> {
    pub fn new(audio_node: Rc<A>, get_param_fn: F) -> Self {
        Self { audio_node, get_param_fn }
    }

    pub fn get_param(&self) -> &AudioParam {
        (self.get_param_fn)(self.audio_node.as_ref())
    }
}
```

Anyway, maybe I'm only making things more complicated. Please let me know if I can help anywhere!

**Comment by b-ma** [comment]
Hey, many thanks for having a look! 

> as I understand your issue is probably that you need to return an owned object representing the AudioParam but the web-audio-api-rs only allows to retrieve an &AudioParam reference from an AudioNode? The ParamGetter acts as an owned object delegating access to the param?

Yes that was exactly the problem.

I will try to implement you solution, that's exactly what I the kind of thing I wanted to do but didn't manage to write properly... as it would prevent the `AudioParam` to import the whole world I really think that's an improvement :)

> Please let me know if I can help anywhere!

Thanks, sure I will!

(`napi-rs` just release their v2, I will try to see where it gets us...)

**Comment by b-ma** [comment]
Hey, things start to get a bit serious here too :) I managed to reproduce both the `many-oscillator-with-env` and the `granular-scrub` examples, so I have the impression most of the very important points are tackled (maybe not in the most elegant way, but that works).

I tried to implement the solution you proposed, but fall again in some kind of `dyn` + `lifetime` trap, which is something that is still way beyond my rust skills... 

So if you have some ideas on how to simply clean the thing I would be happy to have your help there (the two ugly / verbose parts are the `AudioParamGetter` and the `AudioNode::connect`). However, as I also really think these parts can be quite easily generated from the IDL (as most of the `AudioNode`s), maybe this doesn't even worth the hassle to try to clean that, would welcome your thoughts on that.

(I will invite you as collaborator, does not mean I expect you to contribute actively but I think it make sens somehow)
